### ISSUE TITLE
Investigate occurrences of time entries updates not being propagated to the calendar

### ISSUE THREAD

**Issue by daividssilverio** [toggl/mobileapp#4730]
This is most likely caused by some thread marshalling problem and exceptions being swallowed. 

Not easy to track. One case was documented here: https://github.com/toggl/mobileapp/issues/4599#issuecomment-474357310

**Comment by daividssilverio** [comment]
Oh well, there's an exception being thrown when replacing the calendar items collection.

`System.Exception: Realm accessed from incorrect thread.`

It happens on: 
https://github.com/toggl/mobileapp/blob/d22ed9880b00454164ff07d24e28bea179e43dcf/Toggl.Foundation.MvvmCross/ViewModels/Calendar/CalendarViewModel.cs#L223-L232

More specifically on Line 231.

I thought it would've been enough to fix it to change the subscription thread and downstream threads by adding 

```
 .SubscribeOn(schedulerProvider.BackgroundScheduler) 
 .ObserveOn(schedulerProvider.BackgroundScheduler)
```

to the stream, but it doesn't fix the problem, it seems to only soften it.

I'm not sure anymore what could be causing it or how to fix it.
Maybe @juanlaube (calendar expert), @simonrozsival (realm) or @Odrakir (rx) could help us with this.

**Comment by daividssilverio** [comment]
Oh yeah, one thing that did help, at least during my tests was to handle the exception and just do nothing with it. The calendar self recovered after another event in the stream.

As it currently is, the stream gets disposed and calendar receives no more updates.
@amulware can we have that as an workaround in case we can't figure out how to fix this for good?

**Comment by amulware** [comment]
I'd much rather fix this properly. We really shouldn't have realm thread errors, it shouldn't be that hard to convert from realm model to threadsafe model before switching threads.

**Comment by amulware** [comment]
But if this is a blocker, poke me later and we can discuss it.

**Comment by juanlaube** [comment]
This is an interesting issue. I was able to reproduce this twice, at some point the calendar view doesn't update anymore, however I can't find why this happens. Adding an exception breakpoint in Rider didn't help also.
Ideas? 🤔

**Comment by daividssilverio** [comment]
The exception happens here:
https://github.com/toggl/mobileapp/blob/d22ed9880b00454164ff07d24e28bea179e43dcf/Toggl.Foundation.MvvmCross/ViewModels/Calendar/CalendarViewModel.cs#L223-L232

More specifically
```
 after this finishes ->  .SelectMany(_ => reloadData()) 
 and the exception is thrown from here ->   .Subscribe(CalendarItems.ReplaceWith) 
```

You don't see it very easily actually, I've got to put `.Debug` everywhere to actually find this.
After that, I replaced:
```
.Subscribe(CalendarItems.ReplaceWith) 
```

with
```
.Subscribe(items => {
  try {  CalendarItems.ReplaceWith(items); }
  catch (Exception ex) { print error }
  }
) 
```

to see the stack trace.

**Comment by daividssilverio** [comment]
@amulware, the conversion from realm to thread safe should already be happening in the data source.

I'd feel uncomfortable to release it as it is because users always find those edge cases, but also, it's quite hard to replicate.

If I no one else can take a look and maybe help finding the root of the problem, I'd go with the workaround. 

@juanlaube already took a look and can say more.

**Comment by juanlaube** [comment]
I tried to subscribe on different threads but everything makes this worse. I'm not sure the problem is related to the calendar, but could be something deeper in the interactor or data source, we already know of some threading exceptions with realm and this could be related.

About how hard is to reproduce this, I'd say it's not hard at all, I can reproduce by creating multiple time entries from a calendar event (just by tapping on the event really fast), it's an uncommon scenario though, I had to create 20 time entries to make it throw.

**Comment by amulware** [comment]
The datasource/interactors shouldn't be that complicated, right?

This is where the conversion should happen:
https://github.com/toggl/mobileapp/blob/8e79b69545e62714ac8af8a6a3c04397cf43db5d/Toggl.Foundation/DataSources/DataSource.cs#L41-L42

Does it? Doesn't it? On the same thread as the realm call? I don't know. Try sticking a `ToList` at the end here and see what happens?

Looks like the conversion `Select` there is in the same observable chain that gets returned ultimately? I.e. none of that code is called until you subscribe in the `CalendarViewModel`.


RE:
```
.SubscribeOn(schedulerProvider.BackgroundScheduler)
.ObserveOn(schedulerProvider.BackgroundScheduler)
```
Are you sure this ensures it'll all work on the same thread? Not just 'some background thread'?
(Though, using `SubscribeOn` in general seems to make sense to me to ensure the whole operation doesn't happen on the ui-thread, right?)

Just some thoughts..

**Comment by amulware** [comment]
Actually, I'm not sure that `SubscribeOn` actually does anything - you're already locked into the thread that the `ItemsChanged` came from, and who knows where that is.

**Comment by amulware** [comment]
Darp. I only now saw the PR above. :D

**Comment by juanlaube** [comment]
Closed by #4746
